mailpile/plugins/vcard_face.py

```diff
@@ -0,0 +1,57 @@
+import time
+
+from mailpile.i18n import gettext as _
+from mailpile.mailutils import Email
+from mailpile.plugins import PluginManager
+from mailpile.search import MailIndex
+from mailpile.vcard import MailpileVCard, VCardImporter, VCardLine
+
+
+_plugins = PluginManager(builtin=__file__)
+
+
+class FaceImporter(VCardImporter):
+    """
+    Get avatar from Face header.
+    """
+    FORMAT_NAME = 'Face'
+    FORMAT_DESCRIPTION = _('Get avatar from Face header.')
+    SHORT_NAME = 'face'
+    CONFIG_RULES = {
+        'active': [_('Enable this importer'), bool, True],
+    }
+    VCARD_TS = 'x-face-ts'
+    VCARD_IMG = ''
+
+    def get_vcards(self):
+        mail_index = MailIndex(self.session.config)
+        mail_index.load(self.session)
+        results = []
+
+        for vcard in self.session.config.vcards.values():
+            # Get latest message from vcard email address.
+            from_messages = list(mail_index.search(
+                self.session, ['from:' + vcard.email]).as_set())
+            if not from_messages:
+                continue
+            mail_index.sort_results(self.session, from_messages, 'date')
+            latest_message = Email(mail_index, from_messages.pop()).get_msg()
```

Comment:
You probably want to add a "has:face" keyword, by registering a hook with register_meta_kw_extractor(), see `mailpile/plugins/cryptostate.py` for examples of that.

Basically you'll just check if the parsed message has a Face header and add return `set(["face:has"'])` (yes, backwards, sorry internals are weird) if so - probably about four lines of code. You can then efficiently search for "has:face".

Regarding performance - the importers are invoked periodically and will attempt to synchronize ALL contacts. So if you have 10000 contacts, your current code will attempt to load, parse and evaluate 10000 e-mails. That is a lot of disk seeks and CPU time to accomplish nothing much in the common case where nothing has changed. And no, there's no caches that will help - but this will happily flush any caches we were relying on elsewhere, thus magnifying the performance hit even further. This goes down if we scope things and search for has:face - until everyone starts adding Face: headers to their mail. :-)